lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -265,6 +273,142 @@ getBlocks logTrace logic recoveryHeadersMessage enqueue nodeId tipHeaderHash che
+    -> EnqueueMsg
+    -> NodeId
+    -> HeaderHash
+    -> [HeaderHash]
+    -> ((Word32, Maybe Gauge, Conc.TBQueue StreamEntry) -> IO t)
+    -> IO (Maybe t)
+streamBlocks _        _   _     0            _       _      _         _           _ = return Nothing -- Fallback to batch mode
+streamBlocks logTrace smM logic streamWindow enqueue nodeId tipHeader checkpoints k = do
+    blockChan <- atomically $ Conc.newTBQueue $ fromIntegral streamWindow
+    let wqgM = dhStreamWriteQueue <$> smM
+    fallBack <- atomically $ Conc.newTVar False
+    requestVar <- requestBlocks fallBack blockChan
+    r <- k (streamWindow, wqgM, blockChan) `finally` (atomically $ do
+        status <- Conc.readTVar requestVar
+        case status of
+             OQ.PacketAborted -> pure (pure ())
+             OQ.PacketEnqueued -> do
+                 Conc.writeTVar requestVar OQ.PacketAborted
+                 pure (pure ())
+             OQ.PacketDequeued asyncIO -> pure (cancel asyncIO))
+    r' <- atomically $ Conc.readTVar fallBack
+    if r' then pure Nothing
+          else pure $ Just r
+  where
+
+    writeStreamEnd :: Conc.TBQueue StreamEntry -> IO ()
+    writeStreamEnd blockChan = atomically $ Conc.writeTBQueue blockChan StreamEnd
+
+    mkStreamStart :: [HeaderHash] -> HeaderHash -> MsgStream
+    mkStreamStart chain wantedBlock =
+        MsgStart $ MsgStreamStart
+        { mssFrom = chain
+        , mssTo = wantedBlock
+        , mssWindow = streamWindow
+        }
+
+    requestBlocks :: Conc.TVar Bool -> Conc.TBQueue StreamEntry -> IO (Conc.TVar (OQ.PacketStatus ()))
+    requestBlocks fallBack blockChan = do
+        convMap <- enqueue (MsgRequestBlocks (S.singleton nodeId))
+                            (\_ _ -> (Conversation $ \it -> requestBlocksConversation blockChan it `finally` writeStreamEnd blockChan) :|
+                                      [(Conversation $ \it -> requestBatch fallBack blockChan it `finally` writeStreamEnd blockChan)]
+                                     )
+        case M.lookup nodeId convMap of
+            Just tvar -> pure tvar
+            -- FIXME shouldn't have to deal with this.
+            -- One possible solution: do the block request in response to an
+            -- unsolicited header, so that's it's all done in one conversation,
+            -- and so there's no need to even track the 'nodeId'.
+            Nothing   -> throwM $ DialogUnexpected $ "requestBlocks did not contact given peer"
+
+    requestBatch
+        :: Conc.TVar Bool
+        -> Conc.TBQueue StreamEntry
+        -> ConversationActions MsgGetBlocks MsgBlock
+        -> IO ()
+    requestBatch fallBack blockChan _ = do
+        -- The peer doesn't support streaming, we need to fall back to batching but
+        -- the current conversation is unusable since there is no way for us to learn
+        -- which blocks we shall fetch.
+        atomically $ writeTVar fallBack True
+        atomically $ Conc.writeTBQueue blockChan StreamEnd
```

Comment:
`StreamEnd` will be already written by the finally handler above.

Reply:
Will fix.